PR-12687 Use built-in DMA for enqueueCopyBuffer in edge devices

xdavidz opened the pull request:
This is the zocl ARM64 DMA for copy buffer. Performance test results show DMA memcpy gets 25x to 71x better performance than CPU memcpy.

Function Spec and test result: http://confluence.xilinx.com/display/XSW/FS+-+built-in+DMA+for+enqueueCopyBuffer

Solution:
On ARM64 ZCU+ devices, the built-in DMA is always available, thus we should use dma_memcpy for copy buffer.

gbuildx commented:
Build Passed!

larry9523 commented:
After offline discussion, we all agree to implement the copy BO through ERT command instead of ioctl.

maxzhen commented:
Please also take care of removing xclCopyBO() API.

gbuildx commented:
Build Passed!

xdavidz commented:
> After offline discussion, we all agree to implement the copy BO through ERT command instead of ioctl.

It takes longer than I expected because we don't want to pass in scheduler headers into zocl_bo.c. The solution would be separating scheduler, bo and dma to 3 different files with different scope of header files. The scheduler will register itself (the sched_cmd) with a callback function. The dma will handle dma prepare, submit and completion. The bo will only handle address. ETA: today.

xdavidz commented:
> After offline discussion, we all agree to implement the copy BO through ERT command instead of ioctl.

Switched to ERT command and removed xclCopyBO. Please review.

xdavidz commented:
> Please also take care of removing xclCopyBO() API.

I would take care of this separately, need to understand how to deprecate published API. Will get it done asap.

gbuildx commented:
Build Passed!

xdavidz commented:
I could not find appropriate place for preserving the Test code, so just uploaded here.
https://github.com/xdavidz/temp/tree/master/built-in-dma

mamin506 commented on src/runtime_src/core/edge/drm/zocl/sched_exec.c:
```diff
@@ -1384,6 +1394,29 @@ notify_host(struct sched_cmd *cmd)
+/**
+ * Note: zocl copy bo will use built-in dma without
+ * using any real ERT CU kernel, we will need to increase
+ * the poll count to weak up scheduler when dma is done
```
'wake up'?

larry9523 commented on src/runtime_src/core/edge/drm/zocl/zocl_drv.c:
```diff
@@ -676,6 +677,19 @@ static int zocl_drm_platform_probe(struct platform_device *pdev)
+			ret = -EBUSY;
+			goto err0;
+		}
+	}
```
And at that time, if we don't have a DMA channel, we can simplly fail the command and fall back to memcpy

larry9523 replied:
One thing is missing is ps_ert_submit/ps_ert_query support. But we can leave them for the next PR.

xdavidz replied:
> I think this might be a little bit too aggressive. If we fail to request a DMA channel, do we have to fail to driver probe? Can we move this to the first time we see a copy bo command? This should be a one time thing and should not have impact on performance.

Agree. Using singleton instance for the dma channel and only request if when copy_bo is called first time. I don't think we need a lock here since the scheduler will loop commands one by one.

larry9523 commented on src/runtime_src/core/edge/drm/zocl/sched_exec.c:
```diff
@@ -2210,6 +2304,21 @@ penguin_submit(struct sched_cmd *cmd)
+		}
+	}
+
+
```
delete one blank line?

xdavidz replied:
> delete one blank line?

Fixed.

larry9523 commented:
One thing is missing is ps_ert_submit/ps_ert_query support. But we can leave them for the next PR.

xdavidz commented:
> One thing is missing is ps_ert_submit/ps_ert_query support. But we can leave them for the next PR.

Agree, we need test environment to verify ps_ert works well. We could get it done with Versal platform support.

xdavidz commented on src/runtime_src/core/edge/drm/zocl/sched_exec.c:
```diff
@@ -2388,6 +2497,25 @@ static struct sched_ops ps_ert_ops = {
+	scmd->arg = filp;
+	return 0;
+}
+
```
> It is not very straight forward what this function is doing by its name. I suggest you add some comments. Something like
> 
> 1. It bypass any command opcode except ERT_START_COPYBY
> 2. Reture 1 means the command is valid. ( I think returen true and false would be more proper.
> 3. Use scmd->arg to save filp is not what that field is defined for. So we'd better say on MPSoC, DMA engine is not a real HLS CU and arg is not needed. That is why we can reuse it.

Fixed. 1) add comments; 2) switch to bool and adjust the logic - true means good; 3) add detailed explanation for using arg as private data.

xdavidz commented:
> Please also take care of removing xclCopyBO() API.

@maxzhen @stsoe We discussed with Sonal, will keep those xclCopyBOs but make them more efficient and accurate by 1) allocating/mapping ert BO proactively; 2) document expected behavior and programming guide in ert.h; Should be handled in separate CR .

gbuildx commented:
Build Passed!

xdavidz commented on src/runtime_src/core/edge/drm/zocl/sched_exec.c:
```diff
@@ -2173,25 +2219,97 @@ static void
+	if ((err = zocl_dma_channel_instance(dma_handle, zdev)) != 0)
+		return err;
+
+	// We must set up callback for async dma operations
```
fixed

xdavidz commented on src/runtime_src/core/edge/drm/zocl/sched_exec.c:
```diff
@@ -2173,25 +2219,97 @@ static void
+	};
+	int err = 0;
+
+	// Get single dma channel instance
```
fixed

xdavidz commented on src/runtime_src/core/edge/drm/zocl/sched_exec.c:
```diff
@@ -2173,25 +2219,97 @@ static void
+	dma_cap_mask_t dma_mask;
+
+	if (!dma_handle->dma_chan && ZOCL_PLATFORM_ARM64) {
+		// If zdev_dma_chan is NULL, we haven't initialized it yet.
```
fixed

gbuildx commented:
Build Passed!
